dotCMS/src/main/java/com/dotmarketing/business/web/HostWebAPIImpl.java

```diff
@@ -52,49 +57,97 @@ public Host getHost(HttpServletRequest request)  {
 
-        Host host = null;
-        final HttpSession session   = request.getSession(false);
         final UserWebAPI userWebAPI = WebAPILocator.getUserWebAPI();
-        final User systemUser       = userWebAPI.getSystemUser();
-        final boolean respectFrontendRoles = !userWebAPI.isLoggedToBackend(request);
-        final String pageHostId = request.getParameter("host_id");
+        final User user       = userParam != null ? userParam : userWebAPI.getSystemUser();
+        final boolean respectAnonPerms = user == null || !user.isBackendUser();
 
-        if (pageHostId != null && mode.isAdmin) {
-            host = find(pageHostId, systemUser, respectFrontendRoles);
-        } else {
-            if (session != null && mode.isAdmin && session.getAttribute(WebKeys.CURRENT_HOST) != null) {
-                host = (Host) session.getAttribute(WebKeys.CURRENT_HOST);
-            } else if (request.getAttribute(WebKeys.CURRENT_HOST) != null) {
-                host = (Host) request.getAttribute(WebKeys.CURRENT_HOST);
-            } else {
-
-                final String serverName = request.getServerName();
-                host = resolveHostName(serverName, systemUser, respectFrontendRoles);
-            }
+        Optional<Host> optionalHost = this.getCurrentHostFromRequest(request, user, respectAnonPerms);
+
+        if (!optionalHost.isPresent() && user.isBackendUser()){
+            optionalHost = this.getCurrentHostFromSession(request, user, respectAnonPerms);
         }
 
+        final Host host = optionalHost.isPresent() ? optionalHost.get() : resolveHostName(request.getServerName(),
+                user, respectAnonPerms);
+
+        checkHostPermission(user, respectAnonPerms, host);
+        storeCurrentHost(request, user, host);
+
+        return host;
+    }
+
+    private void storeCurrentHost(final HttpServletRequest request, final User user, final Host host) {
+        final HttpSession session = request.getSession(false);
+
         request.setAttribute(WebKeys.CURRENT_HOST, host);
-        if (session != null && mode.isAdmin) {
 
+        if (session != null && user.isBackendUser()) {
             session.setAttribute(WebKeys.CURRENT_HOST, host);
         }
+    }
 
-        return host;
+    private void checkHostPermission(User user, boolean respectAnonPerms, Host host) throws DotDataException, DotSecurityException {
+        if(!APILocator.getPermissionAPI().doesUserHavePermission(host, PermissionAPI.PERMISSION_READ, user, respectAnonPerms)){
+            String u = (user != null) ? user.getUserId() : null;
+
+            String message = "User " + u + " does not have permission to host:" + host.getHostname();
+            Logger.error(HostAPIImpl.class, message);
+            throw new DotSecurityException(message);
+        }
     }
 
-	@Override
+    private Optional<Host> getCurrentHostFromSession(final HttpServletRequest request, User user, boolean respectAnonPerms)

```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Parameter 'user' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4812018)